# APRIORI MODELING FOR MARKET BASKET ANALYSIS
One of the algorithm for Market Basket Analysis (MBA).<BR>
Proposed by Agrawal and Srikant in 1994. <BR>
Terminologies that should be known in Apriori is support, confidence, lift and conviction. <br>
Apriori is one of Association machine learning model, so in this model Apriori Algorithm will be used with Association Rule Learning.

# I. Import Libraries
To get started, let's import the libraries.

In [0]:
import pandas as pd
from mlxtend.preprocessing import OnehotTransactions
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from mlxtend.preprocessing import TransactionEncoder

# II. Load Data
Read data to perform Market Basket Analysis modeling by input data as 'data' variable.<br>
Here it is the item

![Item](Item.jpg)

In [62]:
data = pd.read_csv('form.csv')
data.head()

Timestamp                 Name  Item 1  Item 2     Item 3
0  2019/09/17 8:58:22 AM GMT+7  Firdaus Adi Nugroho      HP  Racket      Watch
1  2019/09/17 8:58:24 AM GMT+7               faizah      HP  Camera      Watch
2  2019/09/17 8:58:30 AM GMT+7               andrem   Watch  Camera  Music Pad
3  2019/09/17 8:58:30 AM GMT+7                laili  Camera   Watch      Mouse
4  2019/09/17 8:58:33 AM GMT+7                 Tara      HP   Watch  Music Pad

In [7]:
data.shape

(24, 5)

To perform Apriori Algorithm modeling, we only use data that contain only items. So, we need to drop the 'Timestamp' and 'Name' columns.

In [63]:
datafix=data.drop(['Timestamp', 'Name'], axis=1)
datafix.head()

Item 1  Item 2     Item 3
0      HP  Racket      Watch
1      HP  Camera      Watch
2   Watch  Camera  Music Pad
3  Camera   Watch      Mouse
4      HP   Watch  Music Pad

In [32]:
datafix.tail()

Item 1  Item 2  Item 3
19       Soap     Bag  Guitar
20     Router     Bag     NaN
21      Watch  Racket    Soap
22  Music Pad    Soap   Watch
23      Mouse  Camera    Soap

# III. Convert to List
In Apriori Algorithm, we use data in **list** form. So, in this case, dataframe need to be converted to list as preprocessing step before modeling.

In [64]:
datafixx = datafix[['Item 1','Item 2', 'Item 3']].values.tolist() #values.tolist is used to change dataframe to list
datafixx

[['HP', 'Racket', 'Watch'],
 ['HP', 'Camera', 'Watch'],
 ['Watch', 'Camera', 'Music Pad'],
 ['Camera', 'Watch', 'Mouse'],
 ['HP', 'Watch', 'Music Pad'],
 ['Watch', 'Racket', 'Camera'],
 ['HP', 'Camera', 'Watch'],
 ['Watch', 'Camera', 'Music Pad'],
 ['Racket', 'Soap', 'Guitar'],
 ['Racket', 'Camera', 'Guitar'],
 ['Camera', 'Bag', nan],
 ['Music Pad', 'Guitar', 'Camera'],
 ['Camera', 'Watch', nan],
 ['Guitar', 'Camera', 'Music Pad'],
 ['Camera', 'Watch', 'Music Pad'],
 ['Camera', 'Racket', 'Guitar'],
 ['Guitar', 'Camera', 'Watch'],
 ['Guitar', 'Watch', nan],
 ['Camera', 'Watch', nan],
 ['Soap', 'Bag', 'Guitar'],
 ['Router', 'Bag', nan],
 ['Watch', 'Racket', 'Soap'],
 ['Music Pad', 'Soap', 'Watch'],
 ['Mouse', 'Camera', 'Soap']]

# IV. Modeling by Apriori Algorithm
Apriori algorithm is a process used to find frequent-itemset by doing iteration on dataset. Frequent-itemset is an indication of itemset that has a frequency of occurrence more than threshold that has been determined. <br>
Association Rule Learning is data mining techniques to find associative rules between a combination of items. <br>
Terminologies in Association Rule Learning :
  1. Support is an indication of how frequently the itemset   appears in the dataset. 
  2. Confidence is an indication of how often the rule has     been found to be true.
  3. Lift is the ratio of the observed support to that         expected if X and Y were independent.
  4. Conviction is the ratio of expected frequency that X     occurs without Y (rule makes an incorrect prediction).
  
Meanwhile, itemsets are consist of two combination, which is X is called antecedent or left-hand-side (LHS) and Y is called consequent or right-hand-side (RHS).

## Transaction Encoder
In the data, there are some 'nan' or 'null'so we need to remove the 'nan' columns first so the data can be processed in machine algorithm.<br>

Then, processing with **TransactionEncoder** to Encodes database transaction data in form of a Python list of lists into a NumPy array for the sake of memory efficiency when working with large datasets such as doing Market Basket Analysis.

Using and **TransactionEncoder** object, we can transform dataset into an array format suitable for typical machine learning APIs. Via the fit method, the **TransactionEncoder** learns the unique labels in the dataset, and via the transform method, it transforms the input dataset (a Python list of lists) into a one-hot encoded boolean array.

In [73]:
datafixx = [['HP', 'Racket', 'Watch'],
           ['HP', 'Camera', 'Watch'],
           ['Watch', 'Camera', 'Music Pad'],
           ['Camera', 'Watch', 'Mouse'],
           ['HP', 'Watch', 'Music Pad'],
           ['Watch', 'Racket', 'Camera'],
           ['HP', 'Camera', 'Watch'],
           ['Watch', 'Camera', 'Music Pad'],
           ['Racket', 'Soap', 'Guitar'],
           ['Racket', 'Camera', 'Guitar'],
           ['Camera', 'Bag'],
           ['Music Pad', 'Guitar', 'Camera'],
           ['Camera', 'Watch'],
           ['Guitar', 'Camera', 'Music Pad'],
           ['Camera', 'Watch', 'Music Pad'],
           ['Camera', 'Racket', 'Guitar'],
           ['Guitar', 'Camera', 'Watch'],
           ['Guitar', 'Watch'],
           ['Camera', 'Watch'],
           ['Soap', 'Bag', 'Guitar'],
           ['Router', 'Bag'],
           ['Watch', 'Racket', 'Soap'],
           ['Music Pad', 'Soap', 'Watch'],
           ['Mouse', 'Camera', 'Soap']]
te = TransactionEncoder()
te_ary = te.fit(datafixx).transform(datafixx)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

Bag  Camera  Guitar     HP  ...  Racket  Router   Soap  Watch
0   False   False   False   True  ...    True   False  False   True
1   False    True   False   True  ...   False   False  False   True
2   False    True   False  False  ...   False   False  False   True
3   False    True   False  False  ...   False   False  False   True
4   False   False   False   True  ...   False   False  False   True
5   False    True   False  False  ...    True   False  False   True
6   False    True   False   True  ...   False   False  False   True
7   False    True   False  False  ...   False   False  False   True
8   False   False    True  False  ...    True   False   True  False
9   False    True    True  False  ...    True   False  False  False
10   True    True   False  False  ...   False   False  False  False
11  False    True    True  False  ...   False   False  False  False
12  False    True   False  False  ...   False   False  False   True
13  False    True    True  False  ...   False   False  False  False
14  False    True   False  False  ...   False   False  False   True
15  False    True    True  False  ...    True   False  False  False
16  False    True    True  False  ...   False   False  False   True
17  False   False    True  False  ...   False   False  False   True
18  False    True   False  False  ...   False   False  False   True
19   True   False    True  False  ...   False   False   True  False
20   True   False   False  False  ...   False    True  False  False
21  False   False   False  False  ...    True   False   True   True
22  False   False   False  False  ...   False   False   True   True
23  False    True   False  False  ...   False   False   True  False

[24 rows x 10 columns]

In [81]:
frequent_itemsets = apriori(df, min_support=0.05, use_colnames=True)
frequent_itemsets

support                     itemsets
0   0.125000                        (Bag)
1   0.666667                     (Camera)
2   0.333333                     (Guitar)
3   0.166667                         (HP)
4   0.083333                      (Mouse)
5   0.291667                  (Music Pad)
6   0.250000                     (Racket)
7   0.208333                       (Soap)
8   0.625000                      (Watch)
9   0.208333             (Camera, Guitar)
10  0.083333                 (Camera, HP)
11  0.083333              (Mouse, Camera)
12  0.208333          (Music Pad, Camera)
13  0.125000             (Racket, Camera)
14  0.416667              (Watch, Camera)
15  0.083333          (Music Pad, Guitar)
16  0.125000             (Racket, Guitar)
17  0.083333               (Guitar, Soap)
18  0.083333              (Watch, Guitar)
19  0.166667                  (Watch, HP)
20  0.208333           (Watch, Music Pad)
21  0.083333               (Racket, Soap)
22  0.125000              (Watch, Racket)
23  0.083333                (Watch, Soap)
24  0.083333  (Music Pad, Camera, Guitar)
25  0.083333     (Racket, Camera, Guitar)
26  0.083333          (Watch, Camera, HP)
27  0.125000   (Watch, Music Pad, Camera)

In [78]:
association_rules(frequent_itemsets, metric="confidence", min_threshold=0.05)
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1.5)
rules["antecedent_len"] = rules["antecedents"].apply(lambda x: len(x))
rules

antecedents          consequents  ...  conviction  antecedent_len
0               (Mouse)             (Camera)  ...         inf               1
1              (Camera)              (Mouse)  ...    1.047619               1
2              (Racket)             (Guitar)  ...    1.333333               1
3              (Guitar)             (Racket)  ...    1.200000               1
4               (Watch)                 (HP)  ...    1.136364               1
5                  (HP)              (Watch)  ...         inf               1
6              (Racket)               (Soap)  ...    1.187500               1
7                (Soap)             (Racket)  ...    1.250000               1
8   (Music Pad, Guitar)             (Camera)  ...         inf               2
9              (Camera)  (Music Pad, Guitar)  ...    1.047619               1
10     (Racket, Camera)             (Guitar)  ...    2.000000               2
11     (Camera, Guitar)             (Racket)  ...    1.250000               2
12             (Racket)     (Camera, Guitar)  ...    1.187500               1
13             (Guitar)     (Racket, Camera)  ...    1.166667               1
14         (Camera, HP)              (Watch)  ...         inf               2
15              (Watch)         (Camera, HP)  ...    1.057692               1

[16 rows x 10 columns]

In [80]:
rules.shape

(16, 10)

# V. Conclusion
With the minimum threshold of parameter:
- support : 0.05
- confidence : 0.05
- lift : 1.5

We can make some conclusion as below:

**From the support result above, it can be seen that**
- The highest support itemsets with length one is :
  1. {Camera} with support 66.7%
  2. {Watch} with support 62.5%
  3. {Guitar} with support 33.3%
  
- The highest support itemsets with length two is :
  1. {Watch, Camera} with support 41.7%
  2. {Camera, Guitar}, {Music Pad, Camera}, {Watch, Music         Pad} with support 20.8%
  3. {Watch, HP} with support 16.7% 
  
- The highest support itemsets with length three is :
  1. {Watch, Music Pad, Camera} with support 12.5%
  2. And the rest of length three item is with support 8.3%
  
So, the top three sold item is Camera, Watch, and Guitar. Meanwhile, the best combination itemsets is {Watch, Camera}, so those combination can be seen as customer transaction behavior and the seller can combine those two items to gain more profit such as increase the production, combine those items as promo package, increase the awareness of other products, etc.

**From the result of rules above, it can be seen that**<br>
The most correct transaction rules is with confidence = 1, those rules contain such as:
- {Mouse} => {Camera}
- {HP} => {Watch}
- {Music Pad, Guitar} => {Camera}
- {Camera, HP} => {Watch}

which means that for 100% of the transactions containing **{X}** as **antecedent** and **{Y}** as **consequent** the rule is correct or 100% of the times a customer buys **{X}**, **{Y}** will be bought as well.

